typing/typemod.ml

```diff
@@ -1796,34 +1796,38 @@ let check_recmodule_inclusion env bindings =
 
 (* Helper for unpack *)
 
-let rec package_constraints env loc mty constrs =
+let rec package_constraints_sig env loc sg constrs =
+  List.map
+    (function
+      | Sig_type (id, ({type_params=[]} as td), rs, priv)
+        when List.mem_assoc [Ident.name id] constrs ->
+          let ty = List.assoc [Ident.name id] constrs in
+          Sig_type (id, {td with type_manifest = Some ty}, rs, priv)
+      | Sig_module (id, pres, md, rs, priv) ->
+          let rec aux = function
+            | (m :: ((_ :: _) as l), t) :: rest when m = Ident.name id ->
+                (l, t) :: aux rest
+            | _ :: rest -> aux rest
+            | [] -> []
+          in
+          let md =
+            {md with
+             md_type = package_constraints env loc md.md_type (aux constrs)
+            }
+          in
+          Sig_module (id, pres, md, rs, priv)
+      | item -> item
+    )
+    sg
+
+and package_constraints env loc mty constrs =
   if constrs = [] then mty
-  else let sg = extract_sig env loc mty in
-  let sg' =
-    List.map
-      (function
-        | Sig_type (id, ({type_params=[]} as td), rs, priv)
-          when List.mem_assoc [Ident.name id] constrs ->
-            let ty = List.assoc [Ident.name id] constrs in
-            Sig_type (id, {td with type_manifest = Some ty}, rs, priv)
-        | Sig_module (id, _, md, rs, priv) ->
-            let rec aux = function
-              | (m :: ((_ :: _) as l), t) :: rest when m = Ident.name id ->
-                  (l, t) :: aux rest
-              | _ :: rest -> aux rest
-              | [] -> []
-            in
-            let md =
-              {md with
-               md_type = package_constraints env loc md.md_type (aux constrs)
-              }
-            in
-            Sig_module (id, Mp_present, md, rs, priv)
-        | item -> item
-      )
-      sg
-  in
-  Mty_signature sg'
+  else begin
+    match Mtype.scrape env mty with
+    | Mty_signature sg ->
+        Mty_signature (package_constraints_sig env loc sg constrs)
+    | _ -> mty
```

Comment:
It looks like this case was not allowed originally (in particular functor types).
What is the rationale?

Reply:
I had thought that a specification for `package_constraints` of: "If possible strengthen this module type with these constraints otherwise return the original weaker type" was reasonable, which is why I wrote it this way.

However, the function is used both co- and contra-variantly, so actually it would probably be better to `assert false` in these cases explicitly. I'll make that change.

Reply:
Since the syntax allows you to write that, you should rather raise an error, shouldn't you ?
This was the role of the original `extract_sig`.

Reply:
Ah, I had missed that `modtype_of_package` also gets used for initially typing the package syntax. I was only interested in the use in `package_subtype` and missed the other ones.

I'll fix the function to work properly for those uses and add some test cases -- something that would have broken for this version of the patch.

Reply:
On closer inspection, my first impression was right. `modtype_of_package` is only called on the contents from existing `Tpackage` types -- so anything that goes wrong here is a bug in the type-checker not an error. I've pushed changes that use `assert false` for this case and simplify the rest of the function code so that it well have produce assertion failures rather than errors.